pandas/core/reshape/pivot.py

```diff
@@ -207,26 +207,30 @@ def __internal_pivot_table(
                 to_unstack.append(i)
             else:
                 to_unstack.append(name)
-        table = agged.unstack(to_unstack)
+        table = agged.unstack(to_unstack, fill_value=fill_value)
 
     if not dropna:
         if isinstance(table.index, MultiIndex):
             m = MultiIndex.from_arrays(
                 cartesian_product(table.index.levels), names=table.index.names
             )
-            table = table.reindex(m, axis=0)
+            table = table.reindex(m, axis=0, fill_value=fill_value)
 
         if isinstance(table.columns, MultiIndex):
             m = MultiIndex.from_arrays(
                 cartesian_product(table.columns.levels), names=table.columns.names
             )
-            table = table.reindex(m, axis=1)
+            table = table.reindex(m, axis=1, fill_value=fill_value)
 
     if sort is True and isinstance(table, ABCDataFrame):
         table = table.sort_index(axis=1)
 
     if fill_value is not None:
-        table = table.fillna(fill_value, downcast="infer")
+        table = table.fillna(fill_value)
+        if aggfunc is len and not observed and lib.is_integer(fill_value):
+            # TODO: can we avoid this?  this used to be handled by
+            #  downcast="infer" in fillna
+            table = table.astype(np.int64)
```

Comment:
convert_dtypes? Also I'm thinking we should be recommending convert_dtypes over infer_objects as it will handle cases like these.

Reply:
Ack - never mind. I missed that this was for converting to the nullable/pyarrow dtypes.

Reply:
My edits above probably made this hidden, so friendly ping on this bit:

> What do you think about essentially exposing `maybe_downcast_to_dtype` with `dtype="infer"` as a method on Series/DataFrame? Then we could do (and users could do) something like `ser.downcast()`.

Reply:
My recollection of discussion surrounding #40988 was a "wait and see if anyone complains" before adding a new method.  In practice anything _other_ than "infer" is pretty weird. while a "downcast round-floats-to-ints" or "downcast to smaller-itemsize" can be useful, that has a lot of overlap with pd.to_numeric and im hesitant to add a new method that is really similar to an existing function